### DISCUSSION TITLE
UPBGE 0.3 - Some feature requests for logic bricks

### DISCUSSION THREAD

**Discussion by arch-sketch** [UPBGE/upbge#1496] [Ideas/Features for UPBGE 0.3+]
**Change the way lists are displayed**

Sorry, I dont know the terminology here, but when you press on a drop down list slot in the logic bricks in order to pick an item, the names of the items are displayed differently from how they are displayed elsewhere in blender - long names are cut off at their end, while in blender currently the names are cut off in the middle, and the ends are shown, which is much more useful when using long names to organize your objects in the scene.

![a04](https://user-images.githubusercontent.com/81666035/120216614-a1368400-c20d-11eb-9bad-a3d477e7589b.jpg)

Notice the difference between the drop down list name in blue and the one shown in the red box.

PS: Another request would be to add an object picker for the Message actuator, since it is used frequently and right now you need to be scrolling the names of the objects of the entire scene, instead of being able to manualy pick them as you would with the Edit object actuator.

![a05](https://user-images.githubusercontent.com/81666035/120217259-7b5daf00-c20e-11eb-921d-b262eed7ad42.jpg)

**Comment by arch-sketch** [comment]
**Add the option of using a property as value for Property actuators**

For operations like adding and subtracting, it would be very useful to be able to use a property instead of a typed number as Value, so you can use the current value defined in the property, instead of a fixed one.

Example: I want to add 'ammo extra' to the 'ammo' amount, but 'ammo extra' is defined in another object (needs to be copied) and can change over time. So instead of a number we would set the property itself as a value for the add operation.

![a06](https://user-images.githubusercontent.com/81666035/120218658-739f0a00-c210-11eb-924c-01eed0b3ef5f.jpg)

PS: This could also be useful for other logic bricks, like Property sensors, for several operations, like the values for Intervals, Greater and Less than, Equal, etc.

**Comment by BluePrintRandom** [comment]
"actuator use property"

I actually had this request before - it's a good one,
when coupled with 'sensor set property' - it makes the logic bricks as
strong as nodes

**Reply by arch-sketch** [reply]
Indeed, it would definitely increase their versatility.

**Comment by arch-sketch** [comment]
**Update property name changes automatically inside logic bricks**

When a game property name is changed, it would be best if it could automatically update in any logic bricks using it. Right now it goes red, signalling there's an error, but you have to manually change to the new property in each brick where it is used, which can be time consuming for long logic trees, and also prone to mistakes and/or omissions.  

![property](https://user-images.githubusercontent.com/81666035/120511522-de7f4b00-c3a0-11eb-8117-e94b9a6cd1ba.jpg)

**Comment by arch-sketch** [comment]
**Add Comments "logic bricks"**

This one Im not sure what design would be best (I did make a quick mockup of one possible solution), but it would be very useful to be able to have comments inside a logic tree to document them, for explanations, annotations, visual clarity, etc.
One way this could work would be to be able to add Comments "logic bricks" in the stack, that have a simple text input, for any of the sections (sensors, controllers, actuators). Having them as bricks would allow for visual guidance as well, as they could work as "breaks" visually separating groups of bricks.

![01](https://user-images.githubusercontent.com/81666035/120674496-8876da00-c46a-11eb-84f7-927e62f026f1.jpg)

**Comment by arch-sketch** [comment]
**Allow for the scrolling or zooming of the editor window while holding on the connector link**

Right now when you are in the process of linking between two bricks, you cant zoom or scroll the editor window while holding on the link 'noodle', which makes it very awkward when the logic tree cant fit entirely inside the window area. You are forced to zoom out and have both bricks in view before connecting them, which is a problem when the logic tree is long. It would be very useful to be able to still use the middle mouse button to scrool or zoom while holding the connector link with the left mouse button. Optionally, it would be also nice if the window could auto scroll if the mouse is near the border of the window, while holding.

![logic01](https://user-images.githubusercontent.com/81666035/122830243-75636700-d2be-11eb-8bf8-6db73191b825.jpg)

**Comment by arch-sketch** [comment]
**Use Properties as input values for sensors and actuators**

Elaborating a bit on what BluePrintRandom mentioned and the request above, it would be extremely useful to be able to use the current value of a property as input value for sensors and actuators. Not only in terms of convenience (editting a single value that will aplly to several logic bricks at once), but more importantly to allow for certain operations that require the input values to be read dynamically (in this example, the speed value could be changed during gameplay and the values updated accordingly).

![properties01](https://user-images.githubusercontent.com/81666035/122833710-d2155080-d2c3-11eb-9a15-757c5965e9ca.jpg)

![properties02](https://user-images.githubusercontent.com/81666035/122835202-4e109800-d2c6-11eb-9867-65ac971d8244.jpg)

**Comment by arch-sketch** [comment]
**Make logic brick's names easier to read**

When organizing a logic tree, it becomes very important to name the logic bricks in a descriptive way. This often means that the names get truncated, since the name input box display area is very small. The only way to read them in their entirity is to hover the mouse on top of it and wait a second for the pop up  info window to appear, with its full description. This however is only possible if your logic brick is already maximized. If its minimized, hovering over the name wont invoke the info window. While it would be great in the future if we could improve the design of the logic bricks to allow for more area to display the names (so you can read them on the whole logic tree at a glance), a useful alternative would be to have the popup info window appear when you hover the mouse on top of the names of **minimized** logic bricks (and not only when they are opened). This would solve the inconvenience of having to maximize each brick you want to inspect for its full name.

![names](https://user-images.githubusercontent.com/81666035/123876738-38c4ea80-d912-11eb-808a-a7a82f99eca0.jpg)

**Comment by arch-sketch** [comment]
**Set direction for the mouse movement sensor**

We have a logic brick sensor to detect if mouse movement is true or false, but no way of specifying which direction we want to be detected. It would be great if we can add the same functionality we have for the movement sensor, to the mouse movement sensor.

![01](https://user-images.githubusercontent.com/81666035/126045983-97dadb0b-35f6-4dd9-ae5b-22ffe2b5fef7.jpg)

![01a](https://user-images.githubusercontent.com/81666035/126045985-f6971472-5302-4830-a820-71d75a5eac43.jpg)

So when we choose mouse movement, we can specify the axis, local/global and threshold

**Comment by arch-sketch** [comment]
**Add smoothing to the mouselook actuator**

Right now, the 'vanilla' mouselook offered by the actuator is lacking in terms of smooth movement. It feels rough, the flow is visibly jagged. Although there are ways of adressing this via python, It would be very useful to have a bult in 'smoothing' value added to the mouselook actuator. This would be useful for all beginners and non gaming users (archviz, teachers, etc) where it wouldnt be practical to develop or search online for a script in order to have a smooth looking mouselook. 

![1](https://user-images.githubusercontent.com/81666035/127753319-4d363a2c-fb49-4ca6-a559-5b4afc51442e.jpg)

**Comment by arch-sketch** [comment]
**Add 'blendout' functionality to the action actuator**

Currently there's a lack of a 'blendout' function to allow for an animation playing on top (layer above) of another animation (layer below), to seamlessly blend back to it after being played (the same effect the blendin function does at the start). 
Bones will smoothly transition from one animation to another because of the blendin, but will suddenly snap back into position at the end of the top animation (no blendout).

However, this feature does exist in blender right now, in the NLA editor. 

![01](https://user-images.githubusercontent.com/81666035/127780206-7b5cd2c4-4552-4390-8243-6df87ebeeb22.jpg)

It has the exact expected fuctionality, allowing for smooth blending of key frames at the start and end of an animation in a top layer (with the animation on the layer below).

This seems like a very important feature to have built into the action actuator. I dont know if the existing NLA mechanics can be ported into the logic brick, but maybe it could be a useful reference.

![02](https://user-images.githubusercontent.com/81666035/127780311-05656431-7cd5-43ed-8df0-96fb99e2b57b.jpg)

**Comment by arch-sketch** [comment]
**Add slow parent functionality (and possibly tracking axis) to the Parent Actuator**

Right now, to parent an object to another, we can use both logic bricks (on game start) or the blender native function (before the game starts). With the native mechanics, we can also set the tracking axis and perhaps more importantly, slow parent. But the parent actuator doesnt provide these options, which means we are lacking these mechanics for parenting relations activated inside/during the game. It would be very useful to have them added to the parent actuator as well, if possible.

![01](https://user-images.githubusercontent.com/81666035/131153269-7013b6db-3bcf-42cc-97d0-a4f10b0ca0cd.jpg)

**Comment by BluePrintRandom** [comment]
I believe you would want the track to actuator to handle tracking
another objects rotation or to rotate to face another object.

maybe a  new mode 'lerp to other rotation'

best not mix the nodes purposes.

**Reply by arch-sketch** [reply]
Although that does sound interesting, I just think the in game parenting options are limited right now. You can remove a parent relation, you can set a parent relation (just object option, I believe). You cant set a slow parent with offset, for example. Again, you can do this "offline" in blender, but not when you parent objects in game, using the parent actuator, because it doesnt offer the options.

**Comment by BluePrintRandom** [comment]
yeah, I believe slow parent is a good idea in parent actuator

however I think 'track other objects rotation' with defined up / forward
should be in the edit object actuator like track to is

**Comment by BluePrintRandom** [comment]
this would allow slow parenting to one object, but using the rotation of
another for instance./

**Reply by arch-sketch** [reply]
Indeed, youle made it possible to use blender's constraints in game by using the 'override logic transform priority' function in the object properties, but the drawback is that you cant use other logic bricks at the same time, they are ignored. So you are limited to either using the constraints or using logic on an object, not both. It would be wonderful in the future if the constraints were to be fully integrated into the realtime environment (like the bone constraints are), which would allow for a lot of flexiblility.
